include/hse/hse.h

```diff
@@ -169,6 +169,26 @@ hse_param_get(const char *param, char *buf, size_t buf_sz, size_t *needed_sz);
  * @{
  */
 
+/** @brief Attach a new KVDB to a snapshot/copy of the media class paths from kvdb_home_src.
+ *
+ * @note This function is not thread safe.
+ *
+ * @param kvdb_home_tgt: KVDB home directory for the attach target.
+ * @param kvdb_home_src: KVDB home directory for the attach source.
+ * @param paths: a snapshot/copy of the media class paths from the attach source.
+ *
+ * @remark @p kvdb_home_tgt and kvdb_home_src must not be NULL.
+ * @remark kvdb_home_src must be offline for attach
+ * @remark A snapshot/copy of all the configured media class paths must be provided in paths.
+ *
+ * @returns Error status.
+ */
+hse_err_t
+hse_kvdb_attach(
+    const char *kvdb_home_tgt,
+    const char *kvdb_home_src,
+    const char *paths[HSE_MCLASS_COUNT]);
+
```

Comment:
C11: 6.7.6.3.7: "A declaration of a parameter as ‘‘array of type’’ shall be adjusted to ‘‘qualified pointer to type’’, where the type qualifiers (if any) are those specified within the [ and ] of the array type derivation. If the keyword static also appears within the [ and ] of the array type derivation, then for each call to the function, the value of the corresponding actual argument shall provide access to the first element of an array with at least as many elements as specified by the size expression."

So, I think if you really want help from the compiler you'd add a "static" in front of HSE_MCLASS_COUNT.  Dunno if gcc or clang ignore it...

Reply:
Interesting, I didn't know that. Will experiment a bit on this.

Reply:
clang honors the static qualifier and throws a warning if the caller doesn't pass at least as many elements as specified for this array.

_warning: array argument is too small; contains 2 elements, callee requires at least 3 [-Warray-bounds]_

gcc ignores the static qualifier, however, its later revisions catches this under -Wstringop-overflow. This worked on gcc 11 but not on gcc 8.

_warning: func accessing 24 bytes in a region of size 16 [-Wstringop-overflow=]_

I think a better option here is to pass a pointer to this array (i.e., _const char * (*paths)[HSE_MCLASS_COUNT]_) rather than the array. This avoids pointer decay and the compiler can now enforce that an array of exact size is passed by the caller.

Reply:
Cool!  Well, I don't think you can ever actually pass an array, just the pointer to it.   But in this case the compiler can enforce that you're passing a ptr to an array of the size the caller expects...
IMO the notation makes the code harder to understand :-(

Reply:
What does the compiler do if it can't determine the size of the array you want to pass to it?

Reply:
Right, yes, I didn't mean to say passing the array but wanted to highlight that we can avoid the pointer decay which is where we lose the size.

Agreed that the notation is not friendlier to read and the caller needs to understand it first to know what to pass :-)

Reply:
> What does the compiler do if it can't determine the size of the array you want to pass to it?

No idea, I presume that would be a warning as well?